scipy/stats/_stats_py.py

```diff
@@ -83,51 +83,39 @@
            'brunnermunzel', 'alexandergovern']
 
 
-# This should probably be rewritten to avoid nested TypeErrors in favor of
-# branching based on dtype.
 def _contains_nan(a, nan_policy='propagate', use_summation=True):
-    if not isinstance(a, np.ndarray):
-        use_summation = False  # some array_likes ignore nans (e.g. pandas)
     policies = ['propagate', 'raise', 'omit']
     if nan_policy not in policies:
         raise ValueError("nan_policy must be one of {%s}" %
                          ', '.join("'%s'" % s for s in policies))
-    # only inexact (floating/complexfloating) and object arrays support np.nan
-    if not (np.issubdtype(a.dtype, np.inexact)
-            or np.issubdtype(a.dtype, object)):
-        return False, nan_policy
-    try:
+
+    if np.issubdtype(a.dtype, np.number):
         # The summation method avoids creating a (potentially huge) array.
         # But, it will set contains_nan to True for (e.g.) [-inf, ..., +inf].
         # If this is undesirable, set use_summation to False instead.
         if use_summation:
             with np.errstate(invalid='ignore', over='ignore'):
-                contains_nan = np.isnan(np.sum(a))
+                # we should use build-in sum instead of np.sum because np.sum
+                # ignores nan when the input is pandas series.
+                contains_nan = np.isnan(sum(a.ravel()))
         else:
             contains_nan = np.isnan(a).any()
-    except TypeError:
-        # This can happen when attempting to sum things which are not
-        # numbers (e.g. as in the function `mode`). Try an alternative method:
-        try:
-            contains_nan = np.any(np.isnan(a))
-        except TypeError:
-            try:
-                # This can happen when attempting to check nan with np.isnan
-                # for string array (e.g. as in the function `rankdata`).
-                contains_nan = False
-                for el in a.ravel():
-                    # isnan doesn't work on elements of string arrays
-                    if np.issubdtype(type(el), np.number) and np.isnan(el):
-                        contains_nan = True
-                        break
-            except TypeError:
-                # Don't know what to do. Fall back to omitting nan values and
-                # issue a warning.
-                contains_nan = False
-                nan_policy = 'omit'
-                warnings.warn("The input array could not be properly "
-                              "checked for nan values. nan values "
-                              "will be ignored.", RuntimeWarning)
+    elif np.issubdtype(a.dtype, np.character) or np.issubdtype(a.dtype, object):
```

Comment:
Can character arrays hold NaNs? I think we only need to check object arrays here.

Reply:
Sorry, I forget that np.nan will be converted to "nan" in string array and we should ignore it. I need to fix this PR. Please wait.

Reply:
No problem. Also, can you compare the execution time of the two approaches (`sum` vs `np.sum`) as array size increases? We need to be confident that this change would not introduce a bottleneck into all the stats functions that use it.

Reply:
Sorry again. You are right. The build-in sum is too slow, maybe this is not acceptable.
```
In [35]: data = np.array(nans_10+list(range(1000000))+nans_10)

In [36]: %timeit np.isnan(np.sum(data))
386 µs ± 4.07 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [37]: %timeit np.isnan(sum(data))
76.6 ms ± 4.51 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```